dist/images/ovndb-raft-functions.sh

```diff
@@ -17,24 +17,59 @@ verify-ovsdb-raft() {
   fi
 }
 
-# OVN DB must be up in the first DB node
-# This waits for ovnkube-db-0 POD to come up
-ready_to_join_cluster() {
+bracketify() { case "$1" in *:*) echo "[$1]" ;; *) echo "$1" ;; esac }
+
+# checks if a db pod is part of a current cluster
+db_part_of_cluster() {
+  local pod=${1}
+  local db=${2}
+  local port=${3}
+  echo "Checking if ${pod} is part of cluster"
+  init_ip=$(kubectl --server=${K8S_APISERVER} --token=${k8s_token} --certificate-authority=${K8S_CACERT} \
+    get pod -n ${ovn_kubernetes_namespace} ${pod} -o=jsonpath='{.status.podIP}')
+  if [[ $? != 0 ]]; then
+    echo "Unable to get ${pod} ip "
+    return 1
+  fi
+  echo "Found ${pod} ip: $init_ip"
+  init_ip=$(bracketify $init_ip)
+  target=$(ovn-${db}ctl --no-leader-only --timeout=5 --db=${transport}:${init_ip}:${port} ${ovndb_ctl_ssl_opts} \
```

Comment:
the init_ip was just getting the pod ip afaict. it was only checking if there is a db running with the pod ip as the transport.

Reply:
the init_ip is used for join-cluster. I have seen the case when a raft member stays as candidate and not be able to connect to the leader (it could be result of ovn bug or underlying networking problem). In that case, the new member attempt to join the cluster is not able to be successful. Of course, this is a corner case.

Reply:
the use of init_ip as a global var in this code is something we should clean up, it just makes things confusing. However, that's outside the scope of this PR I think.

As for the --no-leader-only I think @cathy-zhou we don't want to find the leader, we want to see is this pod is already configured to be part of a cluster. We are already only specifying the db url of this pod, so I'm not sure how we would go to the leader anyway. I think using --no-leader-only is correct, otherwise this command on a db node that is not leader will not work.

Reply:
I think @cathy-zhou 's concern is that the init_ip is used further down to join the cluster and it will fail if the remote-addr passed in (using the init_ip variable) is not that of a leader.

Reply:
So I think if we were to do this right, we need to change the db_part_of_cluster to not set init_ip but instead set some local variable and then in case of success, in cluster_exists, set the init_ip to that of the current leader (rather than that of ovnkube-db-0 pod).

Reply:
hmm...that makes sense but I intentionally did that in the original patch. So let me try to remember why.

Reply:
@trozet in your original code, db_part_of_cluster does not take "--no-leader-only" when "list-connection", so it will return the leader's IP and set it as init_ip in case the cluster already exists. otherwise, if the cluster is first created, uses ovnkube-db-0's IP as the init_ip makes sense, as that is the one who creates the clustered DB.

Reply:
actually @cathy-zhou is correct, we shouldn't be using --no-leader-only. The command is expected to fail if we do not hit the leader. Since we do:

```
  for db_pod in $db_pods; do
    if db_part_of_cluster $db_pod $db $port; then
      echo "${db_pod} is part of current cluster with ip: ${init_ip}!"
      return 0
    fi
  done
```
we just keep overwriting the init_ip until we find the leader. If we dont find the leader, we declare there is no cluster and  set init_ip to ovnkube-db-0

Reply:
That's true. That should be sufficient. Thanks @trozet and @cathy-zhou.

Reply:
The problem with this approach is that in the case that for a period of time that the cluster does not have a leader, then the current code would recreate the cluster. That would also be problematic.

Reply:
If we have no reliable way to determine existence of a cluster other than by saving some annotations/metadata, I think it's Ok to recreate the cluster?

Reply:
we could retry this loop a few times. If we dont find the leader on 3 attempts over some period of several seconds, we can pretty much guarantee the cluster is screwed up.

Reply:
I feel like anything we do here is rather heuristic. If we get to this state, it's ok to recreate the cluster.

Reply:
well actually if there is some transient state, the command will take 5 seconds to time out. So the hope is that in that time period a leader gets decided. I think this is OK for now.